core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/preparation/ParcelPreparationActions.java

```diff
@@ -0,0 +1,235 @@
+    public Action<?, ?> clusterUpgradePreparationFailed() {
+        return new AbstractAction<ClusterUpgradePreparationState, ClusterUpgradePreparationStateSelectors, ClusterUpgradeContext,
+                ClusterUpgradePreparationFailureEvent>(ClusterUpgradePreparationFailureEvent.class) {
+
+            @Override
+            protected ClusterUpgradeContext createFlowContext(FlowParameters flowParameters,
+                    StateContext<ClusterUpgradePreparationState, ClusterUpgradePreparationStateSelectors> stateContext,
+                    ClusterUpgradePreparationFailureEvent payload) {
+                StackView stackView = stackService.getViewByIdWithoutAuth(payload.getResourceId());
+                MDCBuilder.buildMdcContext(stackView);
+                Flow flow = getFlow(flowParameters.getFlowId());
+                flow.setFlowFailed(payload.getException());
+                return ClusterUpgradeContext.from(flowParameters, payload);
+            }
+
+            @Override
+            protected void doExecute(ClusterUpgradeContext context, ClusterUpgradePreparationFailureEvent payload, Map<Object, Object> variables) {
+                String errorMessage = payload.getException().getMessage();
+                Long resourceId = payload.getResourceId();
+                LOGGER.debug("Cluster upgrade preparation failed: {}", errorMessage);
+                ResourceEvent preparationFailedResourceEvent = ResourceEvent.CLUSTER_UPGRADE_PREPARATION_FAILED;
+                fireCloudbreakEvent(resourceId, UPDATE_FAILED, preparationFailedResourceEvent, errorMessage);
+                updateStackStatus(resourceId, DetailedStackStatus.AVAILABLE, preparationFailedResourceEvent);
+                sendEvent(context, HANDLED_FAILED_CLUSTER_UPGRADE_PREPARATION_EVENT.event(), payload);
+            }
+
+            @Override
+            protected Object getFailurePayload(ClusterUpgradePreparationFailureEvent payload, Optional<ClusterUpgradeContext> flowContext, Exception ex) {
+                return null;
+            }
+        };
+    }
+
+    private Image getImageFromCatalog(Stack stack, ImageChangeDto imageChangeDto) throws CloudbreakImageNotFoundException, CloudbreakImageCatalogException {
+        return imageCatalogService.getImage(stack.getWorkspace().getId(), imageChangeDto.getImageCatalogUrl(), imageChangeDto.getImageCatalogName(),
+                imageChangeDto.getImageId()).getImage();
+    }
+
+    private Stack getStack(Long resourceId) {
+        return stackService.getByIdWithClusterInTransaction(resourceId);
+    }
+
+    private ClusterUpgradePreparationEvent createClusterUpgradePreparationEvent(String selector, ClusterUpgradePreparationEvent payload) {
+        return new ClusterUpgradePreparationEvent(selector, payload.getResourceId(), payload.getClouderaManagerProducts());
+    }
+
+    private Set<ClouderaManagerProduct> getRequiredProducts(Stack stack, ImageChangeDto imageChangeDto)
```

Comment:
this method should be in the related action's class, or in a separate class as it is not used from other actions